lib/dma.c

```diff
@@ -580,24 +568,38 @@ dma_controller_dirty_page_get(dma_controller_t *dma, vfu_dma_addr_t addr,
         return ERROR_INT(ENOTSUP);
     }
 
-    if (pgsize != dma->dirty_pgsize) {
+    /*
+     * If dirty page logging is not enabled, the requested page size is zero,
+     * or the requested page size is not a power of two, return an error.
+     */
+    if (dma->dirty_pgsize == 0) {
+        vfu_log(dma->vfu_ctx, LOG_ERR, "dirty page logging not enabled");
+        return ERROR_INT(EINVAL);
+    }
+    if (pgsize == 0 || (pgsize & (pgsize - 1)) != 0) {
         vfu_log(dma->vfu_ctx, LOG_ERR, "bad page size %zu", pgsize);
         return ERROR_INT(EINVAL);
     }
 
-    bitmap_size = get_bitmap_size(len, pgsize);
+    bitmap_size = get_bitmap_size(len, dma->dirty_pgsize);
     if (bitmap_size < 0) {
         vfu_log(dma->vfu_ctx, LOG_ERR, "failed to get bitmap size");
         return bitmap_size;
     }
 
+    converted_bitmap_size = get_bitmap_size(len, pgsize);
+    if (converted_bitmap_size < 0) {
+        vfu_log(dma->vfu_ctx, LOG_ERR, "failed to get bitmap size");
```

Comment:
nit: "bad requested bitmap size %d"

Reply:
This wouldn't work because we're calculating the requested bitmap size so there won't be one if the calculation fails?

Reply:
Sorry I mean "bad requested _page size_ %d"